### PR TITLE
[Reformer] Improved memory if input is shorter than chunk length

### LINKED ISSUE #4660: Assert message error in Reformer chunking
# 🐛 Bug

In the function `apply_chunking_to_forward`, an assertion checking the input tensor shape is trying to print the contents of the tensor itself instead of its shape:

https://github.com/huggingface/transformers/blob/b5015a2a0f4ea63035a877f5626cb0c3ce97e25d/src/transformers/modeling_utils.py#L2195

I'm pretty sure that line should be `input_tensors[0].shape[chunk_dim], chunk_size`

### PR DESCRIPTION
**Pull request by patrickvonplaten**
This PR improves memory and speed of Reformer for language generatinon. 

Reformer is based on chunked self attention. This means that for an input length which is not a multiple of the chunk length, the input has to be padded to be a multiple of the chunk length. This is not the case though when the input length is less than the chunk length (happens in language generation). In this case, normal self attention should be applied to save memory. 

Code is updated for both LSH and Local self attention and a test is added.

### CODE REVIEW

File: tests/test_modeling_reformer.py [hunk 1]
```diff
@@ -829,7 +844,7 @@ def test_local_layer_forward_complex(self):
         reformer_output = layer(prev_attn_output=hidden_states, hidden_states=hidden_states, attention_mask=attn_mask,)
         output_slice = reformer_output.hidden_states[0, 0, :5]
         expected_output_slice = torch.tensor(
-            [1.5476, -1.9020, -0.9902, 1.5013, -0.1950], dtype=torch.float, device=torch_device,
+            [1.4750, -2.0235, -0.9743, 1.4463, -0.1269], dtype=torch.float, device=torch_device,
```
Comment (LysandreJik) [inline tests/test_modeling_reformer.py:847]: This change changes the resulting output?

**Review by LysandreJik — APPROVED** [review]
~That's quite a lot of changes~ The github UI makes it look like a lot of changes, glad to see it improves the memory footprint!

File: tests/test_modeling_reformer.py (hunk 1 above)
Comment (patrickvonplaten) [reply tests/test_modeling_reformer.py:850]: Yeah, good that you mention that - I forgot to adapt the other tests. 

What is happening here is that there was an error previously in the tests:
The input hidden states only had a sequence length of 4 and the chunk length is also 4 but `local_num_chunks_before` was set to 1 which would have meant to attend to the previous chunk as well as its own chunk. This is not possible if there is only one chunk. So I changed `local_num_chunkes_before` to 0. This only affected this test because the attention mask in local self attention depends on the function `self._look_adjacent()`, where as the one in LSH self attention does not.

Comment (LysandreJik) [reply tests/test_modeling_reformer.py:850]: Alright, that makes sense! Thanks for the explanation.
